Add built-in mode config file for Survey tool

hwikle-lanl opened the pull request:
Code review checklist:

- [ ] Code is generally sensical and well commented
- [ ] Variable/function names all telegraph their purpose and contents
- [ ] Functions/classes have useful doc strings
- [ ] Function arguments are typed
- [ ] Added/modified unit tests to cover changes.
- [ ] New features have documentation added to the docs.
- [ ] New features and backwards compatibility breaks are noted in the RELEASE.md

hwikle-lanl committed:
add changes to fresh branch

hwikle-lanl committed:
add survey mode config

hwikle-lanl committed:
add minimial changes back in

hwikle-lanl committed:
add missing types

hwikle-lanl committed:
remove trailing whitespace

hwikle-lanl committed:
add hello_world c program to test survey

Paul-Ferrell commented on builtins/modes/survey.yaml:
```diff
@@ -0,0 +1,50 @@
+        {{ flag(survey.enable-kokkos, "--enable-kokkos") }} 
+        {{ flag(survey.enable-mem, "--enable-mem") }} 
+        {{ flag(survey.enable-mem-gotcha, "--enable-mem") }} 
+        --json-only
```
Parse that json with the json result_parser.

Paul-Ferrell committed:
Merge branch 'master' into survey-mode-config-2

Paul-Ferrell committed:
Fixed hanks problems.

hwikle-lanl committed:
Fix bug when error.request is None

hwikle-lanl committed:
Fix error in hello_c.yaml

hwikle-lanl committed:
Merge branch 'develop' into survey-mode-config

hwikle-lanl committed:
Merge branch 'error-request-none' into survey-mode-config

hwikle-lanl committed:
Partially fix Opt and Sopt plugins

Opt and Sopt plugins are necessary for generating command line
options generated from mode config files. Both Opt and Sopt take
an arguments which can be either a string or a list of strings,
which complicates their arg specs. This commit:

- Adds test cases for both Opt and Sopt plugins.
- Introduces a UnionSpec class to support arguments that are
  a union of multiple types.
- Begins to integrate UnionSpec into argument validation code.

Note that argument validation has not been fully implemented for
UnionSpec, which means that the added tests are not yet passing.

hwikle-lanl committed:
Finish fixing Opt and Sopt function plugins

Opt and Sopt plugins now have correct behavior for lists of strings
and pass all associated tests. Because both plugins can take either
a single string or a list of strings, this commit also finishes
implementing the UnionSpec specification (introduced in the previous
commit), which allows for function plugins to have parameters whose
actual arguments can take on one of several types. This specification
object is designed in the same spirit of the recently introduced Opt
specification, and will hopefully be useful for future plugins.

This commit also modifies the argument validation logic for function
plugins to account for the UnionSpec specification and adds unit tests
as well.

hwikle-lanl committed:
Implement spec description for UnionSpec

hwikle-lanl committed:
Add function plugin for flag

This commit adds a corresponding function plugin for the flag function,
implemented previously. The flag function is used to generate UNIX-style
option flags from boolean-like config options. At present, the flag
function accepts a string as its first argument, rather than a boolean,
the justification being that some versions of YAML specify multiple
conventions for boolean values, including true/false, yes/no, and on/off.
This design decision may need to be revisited at a later date.

hwikle-lanl committed:
Fix syntax error

hwikle-lanl committed:
Fix syntax error and mismatched file names

hwikle-lanl committed:
Add tests for Flag plugin

hwikle-lanl committed:
Fix bug in UnionSpec __contains__ method

hwikle-lanl committed:
Replace UnionSpec argspec with MaybeList

Per Paul, UnionSpec gives users too much power. Instead,
replaced it with MaybeList, which can take either a type
or a list of that same type.

hwikle-lanl committed:
Fix unit test for sopt plugin

hwikle-lanl committed:
Progress towards JSON output

hwikle-lanl committed:
More progress towards mode config

hwikle-lanl committed:
Correct list access in mode config

Paul-Ferrell commented on lib/pavilion/expression_functions/base.py:
```diff
@@ -36,6 +36,43 @@ def num(val):
 
     raise RuntimeError("Invalid value '{}' given to num.".format(val))
 
+def flag(val: str, flag_str: str) -> str:
+    """Return a flag string for a boolean variable, if
+    the variable is set, or an empty string otherwise."""
+
+    if val.lower() in {'true', 'on', 'yes'}:
+        return flag_str
+    elif val.lower() in {'false', 'off', 'no'}:
+        return ''
+    else:
+        raise ValueError(f'Could not convert {val} into boolean-like.')
```
Reverse the argument order. 

`flag('-some_opt', some_opt_var)` reads a lot cleaner to me.

Paul-Ferrell commented on lib/pavilion/expression_functions/base.py:
```diff
@@ -36,6 +36,43 @@ def num(val):
+    else:
+        raise ValueError(f'Could not convert {val} into boolean-like.')
+
+def sopt(val: Union[str, List[str]], option_str: str) -> str:
+    """Return an option string for a variable, if it has a non-null value,
+    or an empty string otherwise. For a list, create separate option string
+    for each value in the list."""
+
+    if isinstance(val, list):
+        return " ".join(map(lambda x: f"{option_str}='{x}'", val))
+    else:
+        if val.lower() in {'null', 'none'}:
+            return ''
+        else:
+            return f"{option_str}='{val}'"
```
Same here:
`sopt('--foo', ['a', 'b', 'c'])` reads better than `sopt(['a', 'b', 'c'], '--foo')`

Paul-Ferrell commented on lib/pavilion/expression_functions/core.py:
```diff
@@ -511,3 +511,44 @@ def outliers(values: List[num], names: List[str], limit: float):
+    def sopt(val: Union[str, List[str]], opt_str: str) -> str:
+        return sopt(val, opt_str)
+
+class OptPlugin(CoreFunctionPlugin):
+    """Convert a config option into a UNIX-style option string. If the option is
+    a list, concatenate the elements of the list to form a single option."""
+    def __init__(self):
+        super().__init__(
+            'opt',
+            arg_specs=(MaybeList(str), str, str)
+        )
+
+    @staticmethod
+    def opt(val: Union[str, List[str]], opt_str: str, delimiter: str = ',') -> str:
+        return opt(val, opt_str, delimiter)
```
The same here on the order, and the last argument can be optional `arg_specs=(str, MaybeList(str), Opt(str))`

hwikle-lanl committed:
Add more unit tests for survey mode config

hwikle-lanl committed:
Fix bug in JSON parser

hwikle-lanl committed:
Make all parsers return log

hwikle-lanl committed:
Pass all results tests

hwikle-lanl committed:
Make unit tests wait for test run to finish

hwikle-lanl committed:
Pass style tests

hwikle-lanl committed:
Implement review feedback

hwikle-lanl committed:
Merge develop

hwikle-lanl committed:
Attempt to pass CI tests

hwikle-lanl committed:
Remove all survey-related functionality

Per Paul Ferrell, LANL will no longer provide Survey as a product.
This commit strips out all Survey-related functionality, but
retains several features and bug fixes implemented during development
of the feature. These include:

- A builtin directory, containing a modes subdirectory, presently empty,
which is checked by when resolving tests, and in which future builtin
modes can be placed.

- Expression functions for building command line arguments and options
from lists and Booleans.

- A fix to a bug in the JSON parser which causes it to incorrectly
return an empty dictionary under certain circumstances.

- Other miscellaneous changes and fixes.

hwikle-lanl committed:
Add gitignore to force inclusion of empty directory

Paul-Ferrell committed:
Merge branch 'develop' into survey-mode-config-2
